pkg/ccl/kvccl/kvtenantccl/tenant_upgrades.md

```diff
@@ -0,0 +1,83 @@
+# Upgrading Tenants
+
+What does it even mean to think about upgrading tenants?
+The upgrade protocol for KV nodes has become well-defined
+over the course of 21.1. It involves tight coordination between
+nodes to gain invariants about what features will be used and
+what migrations can and have been run. SQL pods are inherently
+much more ephemeral and less able to perform this coordination.
+Nevertheless, we need some way to migrate data in tenants and we
+need some claims about the features the tenants may end up using.
+
+* Tenants have their own system tables, including settings.
+* System tables and initial data get bootstrapped in CreateTenant.
+    * We should probably write out an initial version into the settings table
+      at that point.
+    * For new tenants what should we initialize this to?
+        * The best that the system tenant can do is its
+          active version.
+    * For pre-existing tenants in the 20.2->21.1 upgrade, cope
+      with the lack of a value gracefully.
+
+## What does the upgrade dance even look like?
+* I assume we want to upgrade the KV servers *before* we upgrade the
+ tenant pods, at least, sometimes.
+* Would we ever want to upgrade the tenant pods before the KV servers?
+   * Probably yes, but not any time soon.
+* One other consideration is the handshake between SQL pods and KV
+ nodes.
+   * Once the KV cluster has its version finalized, new RPC connections
+     from SQL pods with an old cluster version will not be permitted.
+     Nothing will break the old ones.
```

Comment:
We could prevent this via an EveryNode, where we put in a migration that blocks tenants on version <X from connecting. Bit of a yak shave, but it can be done. Since we own the orchestration, also we just shouldn't let it come to that!

Reply:
Heh yeah, the way the `migration.Cluster` worked out, this will end up happening transparently if we find a way to implement it. Pretty happy with how that refactor shook out.